packages/core/src/modules/credentials/protocol/v2/messages/V2OfferCredentialMessage.ts

```diff
@@ -0,0 +1,63 @@
+import { Expose, Type } from 'class-transformer'
+import { Equals, IsArray, IsInstance, IsOptional, IsString, ValidateNested } from 'class-validator'
+
+import { AgentMessage } from '../../../../../agent/AgentMessage'
+import { Attachment } from '../../../../../decorators/attachment/Attachment'
+import { CredentialFormatSpec } from '../../../formats/models/CredentialFormatServiceOptions'
+import { V2CredentialPreview } from '../V2CredentialPreview'
+
+export interface V2OfferCredentialMessageOptions {
+  id?: string
+  formats: CredentialFormatSpec[]
+  offerAttachments: Attachment[]
+  credentialPreview: V2CredentialPreview
+  replacementId?: string
+  comment?: string
+}
+
+export class V2OfferCredentialMessage extends AgentMessage {
+  public constructor(options: V2OfferCredentialMessageOptions) {
+    super()
+    if (options) {
+      this.id = options.id ?? this.generateId()
+      this.comment = options.comment
+      this.formats = options.formats
+      this.credentialPreview = options.credentialPreview
+      this.messageAttachment = options.offerAttachments
+    }
+  }
+
+  @Type(() => CredentialFormatSpec)
+  @ValidateNested()
+  @IsArray()
+  // @IsInstance(CredentialFormatSpec, { each: true }) -> this causes message validation to fail
+  public formats!: CredentialFormatSpec[]
```

Comment:
I think this issue is happening because the value is not always a class. I've seen multiple occurences of the following:

```
    const formats: CredentialFormatSpec = {
      attachId: this.generateId(),
      format: 'hlindy/cred-filter@v2.0',
    }
```

which are not a class instance but rather a plain object. To make it an actual class instance you'd have to do this:

```ts
    const formats = new CredentialFormatSpec({
      attachId: this.generateId(),
      format: 'hlindy/cred-filter@v2.0',
    })
```

And add the following constructor to the `CredentialFormatSpec` class:

```
  public constructor(options: { attachId: string; format: string }) {
    if (options) {
      this.attachId = options.attachId
      this.format = options.format
    }
  }
```

Typescript is fine with only an object that matches the interface, while we lose the decorators in that case... It's an annoying issue